### DISCUSSION TITLE
Collect all properties that should be configurable via GUI or CLI

### DISCUSSION THREAD

**Discussion by jandubois** [rancher-sandbox/rancher-desktop#1924] [Research]
- Host/App configuration
	- supporting utilities
		- add them to the PATH
		- enable shell completion
	- env variables for nerdctl
	- kubeconfig location
	- autostart at login; maybe show/hide window
	- light/dark/auto mode
	- has admin access
	- set kube context to rancher-desktop when the app is started
- Guest configuration
	- cpu and memory settings
	- local hostname definitions for host resolver
	- proxy settings (http, https, no_proxy)
	- networking config (shared/bridged, MTU)
	- volume mounts (/Users, /Volumes, /private, /tmp, /var/folders)
	- ssh agent forwarding
	- data volume
		- location
		- size (show all of: current size on host disk, used and max size inside the guest
		- include in time machine backup?
	- VM kernel parameters (max file handles etc)
	- (USB pass-through for devices; GPU support)
- Container engine configuration
	- containerd vs. dockerd
	- default containerd namespace: default vs k8s.io
	- docker daemon configuration file?
	- Docker subnet (192.168.65.0/24)
	- CNI
	- registries & mirrors
	- image allow list
- Kubernetes configuration
	- enable/disable k8s
	- enable/disable traefik
	- other k3s options
	- ip range
	- service forwarding
	- kubelet [garbage collection](https://kubernetes.io/docs/concepts/architecture/garbage-collection/) settings
- "First-boot" dialog
	- containerd vs. dockerd
	- enable k8s / k8s version
	- configure supporting utilities
	- sudo (vmnet networking, docker socket)

**Comment by mook-as** [comment]
Some of these are platform-specific; we should mark them to be explicit (and ensure that out understanding matches). 🐧 🍎 🍷 maybe?

**Comment by gunamata** [comment]
Not sure which section here covers ports configuration. For example, users should be able to specify port for Dashboard (steve).

**Reply by jandubois** [reply]
> Not sure which section here covers ports configuration. For example, users should be able to specify port for Dashboard (steve).

I don't think users should ever have to configure this. Why would they ever care? steve should be told to pick a free port, and RD needs to pick that up and then use that port.

Users should never be required to configure or debug something that the computer could figure out on its own.

**Reply by gunamata** [reply]
My main motivation for the comment is this Slack discussion, https://rancher-users.slack.com/archives/C0200L1N1MM/p1648572963209689 

yeah, steve using a random available port should work too. We need to make sure steve doesn't pick any of those standard ports that commonly used applications/services depend on. Or by default let steve pick a random available port, but still give an option for users to pick the port if they want to?

**Reply by jandubois** [reply]
> let steve pick a random available port

That is all we need. Why would a user care to use a particular port? They shouldn't even be aware that `steve` exists; it is an implementation detail.

**Reply by gunamata** [reply]
yeah, I don't mean users might want to use a particular port. Just in case steve randomly picks a port they need for some other application they want to run. As long as we use a non standard port range (eg > 50000), the random port approach is fine.

**Comment by chancez** [comment]
Can we add CNI to the list? Eg #1967. I don't know that RD should directly support installing CNis, as that's a rabbit hole but I think being able to start RD without CNI would be great.

**Comment by rak-phillip** [comment]
@scures This discussion provides a great breakdown for some of the restructuring of preferences in the pipeline. I think we should take these categories into consideration to see if we can expand on the proposal made in #1654

**Comment by byjrack** [comment]
Tack on mtu to the networking configs.

**Comment by Vincy2021** [comment]
I am wondering if the proxy settings (http, https, no_proxy) have been done or not? I can't find where to set proxy and it really bothers me.

**Reply by byjrack** [reply]
Not in the UI or settings.json, but there are places you can pass them into the WSL instance or VM 

Mac - ~/Library/Application\ Support/rancher-desktop/lima/_config/override.yaml
env:
  http_proxy: http://myproxy.com:8080/
  https_proxy: http://myproxy.com:8080/

Linux - ~/.local/share/rancher-desktop/lima/_config/override.yaml
... same format as above

Windows (WSL2)
_Uses the WSLENV user envrionment variable to inject Host environment into WSL distro_
In a Host command prompt:
setx http_proxy http://myproxy.com:8080/
setx https_proxy http://myproxy.com:8080/
setx WSLENV http_proxy:https_proxy

**Reply by Vincy2021** [reply]
Thank you for your reply, actually, I tried that before, but it didn't work. And after your reply，I tried again and still failed. 
The only way I have found to get proxy work is to get into the virtual machine and add environment variables and then restart application "dockerd" again, the command is like that:
sudo HTTP_PROXY="http://myproxy.com:8080/" HTTPS_PROXY="http://myproxy.com:8080/" NO_PROXY="*.test.example.com,.example2.com,127.0.0.0/8" /usr/bin/dockerd

**Reply by byjrack** [reply]
Yeah dockerd can be a bit or a challenge when it comes to timing

I had found in non-RD usage that adding them to /etc/environment seems to get them in place for dockerd startup.  You will also need to update rc_env_allow to "*" for the Alpine OpenRC use.

**Reply by Vincy2021** [reply]
I tried to add them to /etc/environment, but it didn't work either. So, I have to get into the virtual machine and restart "dockerd" from time to time, when I need to add a proxy.

**Comment by jack-liew** [comment]
hi:
i built a private http registry  without SSL.
and it pull & push  with --insecure-registries normally.
but i  can not pull private registry and  set it in rancher desktop gui.

**Reply by jandubois** [reply]
Right now the only way to use a private registry with `docker` in Rancher Desktop is to declare it in `DOCKER_OPTS` in `/etc/conf.d/docker` (inside the VM), e.g.

```
DOCKER_OPTS="--insecure-registry=insecure.registry.internal:80"
```
